notebook/serverextensions.py

```diff
@@ -0,0 +1,197 @@
+
+# Copyright (c) Jupyter Development Team.
+# Distributed under the terms of the Modified BSD License.
+
+from __future__ import print_function
+
+
+import sys
+
+from jupyter_core.paths import jupyter_config_path
+from ._version import __version__
+from .nbextensions import (
+    BaseNBExtensionApp, ToggleNBExtensionApp, _get_config_dir, _read_config_data,
+    _write_config_data, _recursive_update
+)
+
+from traitlets.config.manager import BaseJSONConfigManager
+
+# ------------------------------------------------------------------------------
+# Public API
+# ------------------------------------------------------------------------------
+
+
+class ArgumentConflict(ValueError):
+    pass
+
+
+def enable_server_extension_python(package, user=False, sys_prefix=False):
+    """Enable a server extension associated with a Python package."""
+    data = _read_config_data(user=user, sys_prefix=sys_prefix)
+    server_extensions = (
+        data.setdefault("NotebookApp", {})
+        .setdefault("server_extensions", [])
+    )
+    module, server_exts = _get_server_extension_metadata(package)
+    for server_ext in server_exts:
+        require = server_ext['require']
+        if require not in server_extensions:
+            server_extensions.append(require)
+            diff = {'NotebookApp': {'server_extensions': server_extensions}}
+    _recursive_update(data, diff)
+    _write_config_data(data, user=user, sys_prefix=sys_prefix)
+
+
+def disable_server_extension_python(package, user=False, sys_prefix=False):
+    """Disable a server extension associated with a Python package."""
+    data = _read_config_data(user=user, sys_prefix=sys_prefix)
+    server_extensions = (
+        data.setdefault("NotebookApp", {})
+        .setdefault("server_extensions", [])
+    )
+    module, server_exts = _get_server_extension_metadata(package)
+    for server_ext in server_exts:
+        require = server_ext['require']
+        if require in server_extensions:
+            server_extensions.remove(require)
+            diff = {'NotebookApp': {'server_extensions': server_extensions}}
+    _recursive_update(data, diff)
+    _write_config_data(data, user=user, sys_prefix=sys_prefix)
+
+# ----------------------------------------------------------------------
+# Applications
+# ----------------------------------------------------------------------
+
+
+class ToggleServerExtensionApp(ToggleNBExtensionApp):
+
+    name = "jupyter serverextension enable/disable"
+    description = "Enable/disable a server extension using frontend configuration files."
+
+    def _toggle_server_extension(self, require):
+        config_dir = _get_config_dir(user=self.user, sys_prefix=self.sys_prefix)
+        cm = BaseJSONConfigManager(parent=self, config_dir=config_dir)
+        cfg = cm.get("jupyter_notebook_config")
+        server_extensions = (
+            cfg.setdefault("NotebookApp", {})
+            .setdefault("server_extensions", [])
```

Comment:
It would be great to take this opportunity to make `server_extensions` a dict of bools, like the others. For backward compatibility, we can't change the server_extensions dict itself, but now is the perfect time to make a new traitlet to use that's a dict. The existing list can be deprecated, and add to the dict on startup easily enough.